planner/plannertest/cmp.go

```diff
@@ -0,0 +1,114 @@
+		return errors.New("plans do not have same number of roots")
+	}
+
+	for i, wantNode := range want {
+
+		if seen[wantNode] {
+			continue
+		}
+
+		if _, ok := wantNode.(*planner.LogicalPlanNode); !ok {
+			return fmt.Errorf("encountered a non-logical plan node with spec type %T", wantNode.ProcedureSpec())
+		}
+
+		if _, ok := got[i].(*planner.LogicalPlanNode); !ok {
+			return fmt.Errorf("encountered a non-logical plan node with spec type %T", got[i].ProcedureSpec())
+		}
+
+		if !cmp.Equal(wantNode.ProcedureSpec(), got[i].ProcedureSpec(), semantictest.CmpOptions...) {
+			return fmt.Errorf("logical plan nodes not equal -want/+got %s", cmp.Diff(
+				wantNode.ProcedureSpec(), got[i].ProcedureSpec()))
+		}
+
+		// Nodes are equal
+		seen[wantNode] = true
+
+		if err := CompareLogicalPlans(wantNode.Predecessors(), got[i].Predecessors(), seen); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ComparePhysicalPlans recursively compares two logical query plans.
+// It returns an error if it encounters a non-matching node.
+func ComparePhysicalPlans(want, got []planner.PlanNode, seen map[planner.PlanNode]bool) error {
+	// Check that each plan has the same number of roots
+	if len(want) != len(got) {
+		return errors.New("plans do not have same number of roots")
+	}
+
+	for i, wantNode := range want {
+
+		if seen[wantNode] {
+			continue
+		}
+
+		if _, ok := wantNode.(*planner.PhysicalPlanNode); !ok {
+			return fmt.Errorf("encountered a non-physical plan node with spec type %T", wantNode.ProcedureSpec())
+		}
+
+		if _, ok := got[i].(*planner.PhysicalPlanNode); !ok {
+			return fmt.Errorf("encountered a non-physical plan node with spec type %T", got[i].ProcedureSpec())
+		}
+
+		left := wantNode.(*planner.PhysicalPlanNode)
+		right := got[i].(*planner.PhysicalPlanNode)
+
+		if len(left.RequiredAttrs) != len(right.RequiredAttrs) {
+			return errors.New("nodes don't have the same number of required attributes")
+		}
+
+		for i, attr := range left.RequiredAttrs {
+			if !cmp.Equal(attr, right.RequiredAttrs[i]) {
+				return fmt.Errorf("required attribute not equal -want/+got %s", cmp.Diff(
+					attr, right.RequiredAttrs[i]))
+			}
+		}
+
+		if !cmp.Equal(left.OutputAttrs, right.OutputAttrs) {
+			return fmt.Errorf("output attributes not equal -want/+got %s", cmp.Diff(
+				left.OutputAttrs, right.OutputAttrs))
+		}
+
+		if !cmp.Equal(wantNode.ProcedureSpec(), got[i].ProcedureSpec(), semantictest.CmpOptions...) {
+			return fmt.Errorf("logical plan nodes not equal -want/+got %s", cmp.Diff(
+				wantNode.ProcedureSpec(), got[i].ProcedureSpec()))
+		}
+
+		// Nodes are equal
+		seen[wantNode] = true
+
+		if err := ComparePhysicalPlans(wantNode.Predecessors(), got[i].Predecessors(), seen); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// CreateLogicalPlanTree creates a logical plan tree (not a DAG)
+func CreateLogicalPlanTree(spec planner.LogicalProcedureSpec, predecessors ...*planner.LogicalPlanNode) *planner.LogicalPlanNode {
```

Comment:
I see what you mean about creating an expected plan that is a DAG and not a tree.

I'm thinking that the best way to write test cases that lock down translation from Flux text to the logical plan would use a struct like this:

```go
type TestCase struct {
    name string,
    fluxText string,
    logicalNodes []LogicalPlanNode,
    edges [][2]int,
}
```
Tests can then define a slice of logical nodes, and describe edges using offsets into the slice.  It would be up to the method that runs each test case to create the edges.  What do you think?